File: swift LRUCache/LRUCache.swift
//
//  LRUCache.swift
//  Social Photos
//
//  Created by shoulong li on 1/5/16.
//  Copyright © 2016 shoulong li. All rights reserved.
//

import Foundation

class LRUCache<K: Hashable, V> {
    var capacity: Int
    //header next point to the eldest
    //header prev point to the newest
    //it is a circle
    var header: DoubleLinkedListEntry<K, V>!
    var dictionary: Dictionary<K, DoubleLinkedListEntry<K, V>>
    
    init(capacity: Int) {
        self.capacity = capacity
        dictionary = Dictionary<K, DoubleLinkedListEntry<K, V>>(minimumCapacity: capacity)
        header = DoubleLinkedListEntry<K, V>()
        header.prev = header
        header.next = header
    }
    
    func count() -> Int {
        return dictionary.count
    }
    
    subscript(key: K) -> V? {
        get {
            if let entry = dictionary[key] {
                entry.remove()
                entry.addBefore(header)
                return entry.value
            } else {
                return nil
            }
        }
        
        set(value) {
            //remove the entry if exist
            if let entry = dictionary[key] {
                entry.remove()
                dictionary.removeValueForKey(entry.key!)
            }
            if value != nil {
                let newEntry = DoubleLinkedListEntry(key: key, value: value)
                dictionary[key] = newEntry
                newEntry.addBefore(header)
                if dictionary.count > self.capacity {
                    removeEldestEntry()
                }
            }
        }
    }
    
    
    func removeEldestEntry() {
        if header.next === header {
            //empty
        } else {
            dictionary.removeValueForKey(header.next!.key!)
            header.next?.remove()
        }
    }
}

class DoubleLinkedListEntry<K: Hashable, V> {
    var key: K? = nil
    var value: V? = nil
    var prev: DoubleLinkedListEntry<K, V>? = nil
    var next: DoubleLinkedListEntry<K, V>? = nil
    
    init() {
        
    }
    
    init(key: K, value: V?) {
        self.key = key
        self.value = value
    }
    
    //add self before the entry
    func addBefore(entry: DoubleLinkedListEntry) {
        prev = entry.prev
        next = entry
        prev?.next = self
        next?.prev = self
    }
    
    func addAfter(entry: DoubleLinkedListEntry) {
        prev = entry
        next = entry.next
        prev?.next = self
        next?.prev = self
    }
    
    func remove() {
        prev?.next = next
        next?.prev = prev
        prev = nil
        next = nil
    }
}

File: Leetcode/LRUCache.java
/*
public class LRUCache {
    private LinkedHashMap<Integer, Integer> hashMap;
    private int capacity;
    public LRUCache(int capacity) {
        hashMap = new LinkedHashMap<Integer, Integer>(2*capacity, 1.5f, true) {
            @Override
            public boolean removeEldestEntry(Map.Entry<Integer,Integer> eldest ) {
                return size() > LRUCache.this.capacity;
            }
        };
        this.capacity =  capacity;
    }
    
    public int get(int key) {
        Integer value =  hashMap.get(key);
        if (value == null ) {
            return -1;
        }
        return value.intValue();
    }
    
    public void set(int key, int value) {
        hashMap.put(key, value);
    }
}
*/
/*

public class LRUCache {

    private HashMap<Integer, DoubleLinkedListNode<Integer, Integer>>  map;
    private DoubleLinkedListNode<Integer, Integer> header;
    private int capacity;
    public LRUCache(int capacity) {
        map = new HashMap<>();
        this.capacity = capacity;
        header = new DoubleLinkedListNode<>();
        header.next = header;
        header.prev = header;
    }

    public int get(int key) {
        DoubleLinkedListNode<Integer, Integer> entry = map.get(key);
        if (entry != null) {
            entry.remove();
            entry.addAfter(header); //make entry the latest recent visited
            return entry.value;
        } else {
            return -1;
        }
    }

    public void set(int key, int value) {
        DoubleLinkedListNode<Integer, Integer> entry = new DoubleLinkedListNode<>(key, value);
        if (map.containsKey(key)) {
            DoubleLinkedListNode<Integer, Integer> existOne = map.get(key);
            existOne.remove();
            map.remove(key);
        }
        map.put(key, entry);
        entry.addAfter(header);
        if (map.size() > capacity) {
            removeEldestElement();
        }
    }

    public void removeEldestElement() {
        if (header.prev != header) {
            map.remove(header.prev.key);
            header.prev.remove();
        }
    }
}

class DoubleLinkedListNode<K, V> {
    public K key;
    public V value;
    public DoubleLinkedListNode prev;
    public DoubleLinkedListNode next;
    public DoubleLinkedListNode() {
        key = null;
        value = null;
    }

    public DoubleLinkedListNode(K key, V value) {
        this.key = key;
        this.value = value;
    }

     
    //add this to be before of node
     
    public void addBefore(DoubleLinkedListNode node) {
        next = node;
        prev = node.prev;
        prev.next = this;
        next.prev = this;

    }

    
    //add this to be after of node
     
    public void addAfter(DoubleLinkedListNode node) {
        prev = node;
        next = node.next;
        prev.next = this;
        next.prev = this;
    }

    public void remove() {
        prev.next = next;
        next.prev = prev;
        prev = next = null;
    }
} 
*/
public class LRUCache extends LinkedHashMap<Integer, Integer> {
    private int capacity;
    public LRUCache(int capacity) {
        super(2 * capacity, 1.5f, true);
        this.capacity = capacity;
    }
    public int get(int key) {
        Integer value = super.get(key);
        if (value == null) {
            return -1;
        }
        return value.intValue();
    }

    public void set(int key, int value) {
        super.put(key, value);
    }
    @Override
    public boolean removeEldestEntry(Map.Entry<Integer, Integer> entry) {
        return size() > this.capacity;
    }
}
